libuavcan/src/transport/uc_transfer_buffer.cpp

```diff
@@ -96,18 +96,43 @@ void TransferBufferManagerEntry::Block::destroy(Block*& obj, IPoolAllocator& all
     }
 }
 
+// precondition is that target_offset < total_offset + Block::Size
 void TransferBufferManagerEntry::Block::read(uint8_t*& outptr, unsigned target_offset,
                                              unsigned& total_offset, unsigned& left_to_read)
 {
     UAVCAN_ASSERT(outptr);
-    for (unsigned i = 0; (i < Block::Size) && (left_to_read > 0); i++, total_offset++)
+    // total offset == offset at the start of this block
+    const unsigned start_of_block = total_offset;
+    const unsigned end_of_block = total_offset + Block::Size;
+
+    unsigned bytes_to_copy = 0;
+
+    // Case 1 -- start of wanted area is in this Block
+    if (target_offset >= start_of_block && target_offset < end_of_block)
     {
-        if (total_offset >= target_offset)
+        const unsigned bytes_available_at_offset = end_of_block - target_offset;
+        bytes_to_copy = bytes_available_at_offset;
+        if (bytes_to_copy > left_to_read)
         {
-            *outptr++ = data[i];
-            left_to_read--;
+            bytes_to_copy = left_to_read;
         }
+        uint8_t *src = data + (target_offset - start_of_block);
+        ::memcpy(outptr, src, bytes_to_copy);
```

Comment:
Not to be overly paranoid, but should we not cast the result to `void` to keep compliance with MISRA?

I am aware that the codebase has other non-MISRA-compatible features, such as the use of plain native types like `unsigned` instead of deterministic `uint32_t`. Hopefully that will be resolved one day, too.

Reply:
On this point, I'm not yet able to submit coverity builds. Do we have that scan setup to do MISRA when it does work?

Reply:
IIRC last time I checked MISRA-specific checks were not available on the free plan that we're using. So Coverity is configured to run only general checks.

Reply:
What MISRA rule is this a violation of? I see that 5-0-15 of MISRA C++ allows pointer arithmetic when calculating array indices.

Reply:
ah, done.